openhands/runtime/client/runtime.py

```diff
@@ -194,19 +194,31 @@ def _init_container(
                     f'--plugins {" ".join([plugin.name for plugin in plugins])} '
                 )
 
-            network_mode: str | None = None
-            port_mapping: dict[str, int] | None = None
-            if self.config.sandbox.use_host_network:
-                network_mode = 'host'
+            self._port = self._find_available_port()
+            self.api_url = f'http://{self.config.sandbox.api_hostname}:{self._port}'
+
+            use_host_network = self.config.sandbox.use_host_network
+            network_mode: str | None = 'host' if use_host_network else None
+            port_mapping: dict[str, int] | None = (
+                None if use_host_network else {f'{self._port}/tcp': self._port}
+            )
+
+            if use_host_network:
                 logger.warn(
                     'Using host network mode. If you are using MacOS, please make sure you have the latest version of Docker Desktop and enabled host network feature: https://docs.docker.com/network/drivers/host/#docker-desktop'
                 )
-            else:
-                port_mapping = {f'{self._port}/tcp': self._port}
 
-            if mount_dir is not None:
-                volumes = {mount_dir: {'bind': sandbox_workspace_dir, 'mode': 'rw'}}
-                logger.info(f'Mount dir: {sandbox_workspace_dir}')
+            # Combine environment variables
+            environment = {
+                'PORT': str(self._port),
+            }
+            if self.config.debug:
+                environment['DEBUG'] = 'true'
+
+            logger.info(f'Workspace dir: {sandbox_workspace_dir}')
+            if mount_dir is not None and mount_dir is not UndefinedString.UNDEFINED:
```

Comment:
Is it ever possible to be UNDEFINED here?

Reply:
To elaborate: the problem presented itself in the runtime unit tests (tests/runtime). So just by code or accidental omission of setting it in a config, this could happen.

Reply:
I _think_ I see what's happening: https://github.com/All-Hands-AI/OpenHands/blob/0667f7b935a11f8ef3a9e10a5f95159d947af2ec/tests/runtime/conftest.py#L159
Since we made config a locally passed var, and not loaded by default at import, it's now possible to half-load it, like this: create object manually, load from env manually, but never do the other two calls that always happen in a regular environment: load from toml and finalize_config. Skipping toml is fine (it's optional anyway), but skipping finalize breaks things.

I think we should fix the conftest file, because otherwise we're testing behavior that our users never have.

Reply:
Oh, 🦅 eagly 👁️  eye Enyst, thank you! 😄  I'll try the load_app_config next!